File: src/main/java/de/fluxparticle/animation/signal/SignalInterpolatable.java
package de.fluxparticle.animation.signal;

import javafx.animation.Interpolatable;

public class SignalInterpolatable<T extends Interpolatable<T>> extends SignalPath<T> {

    public static class Builder<T extends Interpolatable<T>> extends SignalPath.Builder<T> {

        @Override
        public SignalPath<T> build() {
            return new SignalInterpolatable<>(start, end);
        }

    }

	public SignalInterpolatable(T start, T end) {
		super(start, end);
	}

	@Override
	public T apply(Double t) {
		return start.interpolate(end, t);
	}

}

File: src/main/kotlin/de/fluxparticle/animation/signal/SignalInterpolatable.kt
package de.fluxparticle.animation.signal

import javafx.animation.Interpolatable

class SignalInterpolatable<T : Interpolatable<T>>(start: T, end: T) : SignalPath<T>(start, end) {

    class Builder<T : Interpolatable<T>> : SignalPath.Builder<T>() {

        override fun build(): SignalPath<T> {
            return SignalInterpolatable(start, end)
        }

    }

    override fun apply(t: Double): T {
        return start.interpolate(end, t)
    }

}
